pkg/kubecfg/show.go

```diff
@@ -19,53 +19,188 @@ import (
+	return strings.Join(parts, "")
 }
 
 func (c ShowCmd) Run(apiObjects []*unstructured.Unstructured, out io.Writer) error {
-	switch c.Format {
+	if c.ExportDir != "" {
+		if err := os.MkdirAll(c.ExportDir, 0777); err != nil {
+			return err
+		}
+		empty, err := isDirEmpty(c.ExportDir)
+		if err != nil {
+			return err
+		}
+		if !empty {
+			return fmt.Errorf("export directory %q is not empty", c.ExportDir)
+		}
+	}
+
+	for i, obj := range apiObjects {
+		if err := c.renderObject(i, obj, out); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c ShowCmd) renderObject(idx int, obj *unstructured.Unstructured, out io.Writer) error {
+	if c.ExportDir != "" {
+		name, err := c.formatObjectName(idx, obj)
+		if err != nil {
+			return err
+		}
+		filename := fmt.Sprintf("%s.%s", name, c.OutputFormat)
+		path := filepath.Join(c.ExportDir, filename)
+		if err := os.MkdirAll(filepath.Dir(path), 0777); err != nil {
+			return err
+		}
+		f, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0666)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		out = f
+	}
+
+	switch c.OutputFormat {
 	case "yaml":
-		for _, obj := range apiObjects {
-			fmt.Fprintln(out, "---")
-			// Urgh.  Go via json because we need
-			// to trigger the custom scheme
-			// encoding.
-			buf, err := json.Marshal(obj)
-			if err != nil {
-				return err
-			}
-			o := map[string]interface{}{}
-			if err := json.Unmarshal(buf, &o); err != nil {
-				return err
-			}
-			buf, err = yaml.Marshal(o)
-			if err != nil {
-				return err
-			}
-			out.Write(buf)
+		fmt.Fprintln(out, "---")
+		o, err := k8sToJSONObject(obj)
+		if err != nil {
+			return err
+		}
+		buf, err := yaml.Marshal(o)
+		if err != nil {
+			return err
+		}
+		if _, err := out.Write(buf); err != nil {
+			return err
 		}
 	case "json":
 		enc := json.NewEncoder(out)
 		enc.SetIndent("", "  ")
-		for _, obj := range apiObjects {
-			// TODO: this is not valid framing for JSON
-			if len(apiObjects) > 1 {
-				fmt.Fprintln(out, "---")
-			}
-			if err := enc.Encode(obj); err != nil {
-				return err
-			}
+		// TODO: this is not valid framing for JSON
+		if idx > 0 {
+			fmt.Fprintln(out, "---")
+		}
+		if err := enc.Encode(obj); err != nil {
+			return err
 		}
 	default:
-		return fmt.Errorf("Unknown --format: %s", c.Format)
+		return fmt.Errorf("Unknown --format: %s", c.OutputFormat)
 	}
-
 	return nil
 }
+
+func k8sToJSONObject(obj *unstructured.Unstructured) (map[string]interface{}, error) {
+	// Urgh.  Go via json because we need
+	// to trigger the custom scheme
+	// encoding.
+	//
+	// Otherwise we could just return obj.Object
+	// encoding.
+	buf, err := json.Marshal(obj)
+	if err != nil {
+		return nil, err
+	}
+	o := map[string]interface{}{}
+	if err := json.Unmarshal(buf, &o); err != nil {
+		return nil, err
+	}
+	return o, nil
+}
+
+func (c ShowCmd) formatObjectName(idx int, obj *unstructured.Unstructured) (string, error) {
```

Comment:
I forgot to add a feature I was planning to add: I'd like to add a function that returns the index of the object so users can decide to render filenames like:

```
├── 00-apps-v1beta2.Deployment-default.frontend.yaml
├── 01-apps-v1beta2.Deployment-default.redis-master.yaml
├── 02-apps-v1beta2.Deployment-default.redis-slave.yaml
```

in order to ensure that when they are migrating from a single giant yaml to a directory full of smaller files, the order of the resources stay the same (assuming that the tool that applies them reads the dir in lexicographical order).

Most tooling (e.g. argocd) shouldn't care about the order of the input files, since they'll do reordering themselves, but when migrating a large codebase from giant single yaml files to forest of yaml files, I'd personally be cautious and use this index.

Will make a commit today with the additional function